### DISCUSSION TITLE
Allow editor to show children of custom class, tool scripts that add child nodes on _ready()

### DISCUSSION THREAD

**Discussion by data-overflow** [godotengine/godot-proposals#9238] [Editor]
Hi! So I'm working at an app with Godot that heavily requires the use of UI. Almost all my UI elements require margins, paddings, panels or layout to so make things easier I wanted to create a node that stacks all these functions together into a single node.

I want a node which can do this to make my development time faster and reduce the complexity of my deeply nested scene tree. 

So my approach was to create a script (extends MarginContainer for margin) with the `class_name` and `tool` keyword so that it shows up in the nodes list and allows me to preview the changes live on the editor. On the `_ready()` function, I'm adding a `Panel` and a `MarginContainer` as its child (for padding). The margins work as expected -

![margins](https://github.com/godotengine/godot-proposals/assets/68324620/3d044531-623d-4c8a-9358-b0f627c669df)

Any children I add to this node via the editor makes it its **immediate children** only. But now for the padding to work I need the nodes I add to this to automatically move to be the children of the `MarginContainer` instead. I did this using the following script - 

```gdscript
func _ready():
	create_nodes()
	set_properties()
	connect("child_entered_tree", _on_child_entered_tree)
		
func create_nodes():
	padding_container = MarginContainer.new()
	panel = Panel.new()
	add_child(panel)
	panel.add_child(padding_container)

func set_properties():
	add_theme_constant_override("margin_top", margin_top)
	add_theme_constant_override("margin_bottom", margin_bottom)
	add_theme_constant_override("margin_left", margin_left)
	add_theme_constant_override("margin_right", margin_right)
	padding_container.add_theme_constant_override("margin_top", padding_top)
	padding_container.add_theme_constant_override("margin_bottom", padding_bottom)
	padding_container.add_theme_constant_override("margin_left", padding_left)
	padding_container.add_theme_constant_override("margin_right", padding_right)
	
func _process(delta):
	if Engine.is_editor_hint():
		set_properties()
		
func _on_child_entered_tree(node):
	remove_child(node)
	padding_container.add_child(node)
```

But then I faced a new problem. When I add nodes to my Div (custom node) it immediately becomes the child of the MarginContainer but also makes it inaccessible from the editor.

![glitch](https://github.com/godotengine/godot-proposals/assets/68324620/00188a8d-a6c0-48c7-9d58-e5b08be63352)

**Error Log**
```
Attempt to disconnect a nonexistent connection from 'Label:<Label#4244467947210>'. Signal: 'size_flags_changed', callable: 'Container::queue_sort'.
  Attempt to disconnect a nonexistent connection from 'Label:<Label#4244467947210>'. Signal: 'minimum_size_changed', callable: 'Container::_child_minsize_changed'.
  Attempt to disconnect a nonexistent connection from 'Label:<Label#4244467947210>'. Signal: 'visibility_changed', callable: 'Container::_child_minsize_changed'.
```

**Is it possible to include this feature to show editor added children nodes to tool scripts?**

**Comment by theraot** [comment]
If you set their `owner` after adding them to the scene tree, they should show up in the editor and also be saved with the scene.

About the signals in the error log you show... I do not know.

At least I can tell you that in tool scripts I'm used to check if a signal is connected before conecting it, and the equivalent for disconnection, that is because the signal connections would also get saved.
